checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -176,6 +197,74 @@ public void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type, bool
         }
     }
 
+    @Override
+    public void preProcessClassTree(ClassTree classTree) {
+        super.preProcessClassTree(classTree);
+        Element elt = TreeUtils.elementFromDeclaration(classTree);
+        List<? extends Element> enclosedElts =
+                ElementUtils.enclosingClass(elt).getEnclosedElements();
+        List<FlowExpressions.Receiver> rs = new ArrayList<>();
+        for (Element e : enclosedElts) {
+            if (!ElementUtils.isEffectivelyFinal(e)) {
+                List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(e);
+                for (AnnotatedTypeMirror atm : enclosedTypes) {
+                    List<FlowExpressions.Receiver> atmRs =
+                            UpperBoundUtil.getDependentReceivers(
+                                    atm.getAnnotations(), getPath(declarationFromElement(e)), this);
+                    rs.addAll(atmRs);
+                    for (FlowExpressions.Receiver rec : atmRs) {
+                        dependentAnnotationByReceiver.put(rec, atm);
```

Comment:
These are the expressions allowed on a declared dependent type:
 * final or effectively final variable
 * local variable
 * private field
 * pure method call all of whose arguments (including the receiver expression) are composed only of expressions in this list
 * access of a public final field whose access expression (sometimes called the receiver) is composed only of expressions in this list

So, only expressions for those kinds need to be kept.  And then only those expression that would cause an error to be issued by `UpperBoundUtil#isSideEffected`.  (I think that is just expressions containing a non-final private field access or a method call.)

Also, I don't see in the code where an error is used if a declared type contains an expression not in that list.  (For annotations on variables, it happens in UBVisitor#visitVariable.)